### DISCUSSION TITLE
Combine attributes_table and table_for

### DISCUSSION THREAD

**Discussion by seanlinsley** [activeadmin/activeadmin#7865] [General]
As of #2544 you can now have an attributes table with multiple entries, like this:

``` ruby
attributes_table_for collection do
  row :id
  row :name
end
```

![](https://f.cloud.github.com/assets/25094/1286164/3e53d99a-2fc7-11e3-976a-0ba84e627969.png)

But we want to move it further; combining `attributes_table_for` and `table_for` into the same simple `table_for` method. In #2544 we discussed multiple approaches at a DSL, but eventually settled on using the `row` & `column` methods to distinguish how the table would be rendered. So:

``` ruby
# A normal table
table_for collection do
  column :foo
  column :bar
end
# A flipped table, displayed like the current attributes table
table_for collection do
  row :foo
  row :bar
end
# This would raise an error
table_for collection do
  column :foo
  row :bar
end
```

As well, both table types would A) accept a single record and B) have a plural shortcuts to easily create multiple columns/rows with a single line of Ruby.

``` ruby
table_for User.last do
  column :id
  column :name
  # or:
  row :id
  row :name
end

table_for User.limit(10) do
  columns :id, :name, :email
  # or:
  rows :id, :name, :email
end
```

**Comment by seanlinsley** [comment]
Looking over the code... how are we going to deal with sortable tables? Are we just going to disallow that feature for transposed tables?

**Comment by shekibobo** [comment]
This works with single records + columns, right? I ask because that case wasn't mentioned in any of the examples above.

**Comment by seanlinsley** [comment]
Yeah, of course. Hence the sentence:

> As well, **both table types** would A) accept a single record and B) have a plural shortcuts to easily create multiple columns/rows with a single line of Ruby.

I'll go ahead and update the example code, though.

**Comment by shekibobo** [comment]
Ahh, I just didn't interpret that right.

Might be clearer to say (if this is going into the docs):

"Both table types would A) accept either a single record or a collection, and B) have plural shortcuts to easily create multiple columns/rows with a single line of Ruby."

I don't mean to be nitpicky, but I had trouble parsing the full meaning of that sentence.

What are your feelings on the transposed sortability? I'm not really sure how I feel. It almost feels to me like it would be cool to be able to manually reorder them, like drag-and-drop, but automatic sorting... idunno. Can these tables be paginated? Is it replacing `index_table`?

**Comment by seanlinsley** [comment]
BTW single-record support for `table_for` was added in 9481d57f7de1a4db9aaf7e2d213b94695788b322

**Comment by seanlinsley** [comment]
> What are your feelings on the transposed sortability? ... Is it replacing index_table?

Right now the code's confusing because `table_for` stores whether or not it thinks itself is sortable, but it doesn't do anything with it. It's higher up in the chain in the index view code that sorting is actually applied. This question so far has blocked me from implementing the rest of this spec.

I'm inclined to say you'd have `sortable_table_for` as a superset of `table_for`, but handles sorting. I'm not yet sure whether it makes sense for a transposed table to be sortable.

The added benefit to `sortable_table_for` is that you could inject a sortable table in non-index pages.

**Comment by shekibobo** [comment]
I like that idea. I really don't think I see much value in sortable transposed tables. I figure you'll only have a few records displayed that way in each table, since larger datasets tend to be better represented in standard table form. And that format of table would tend more toward controlling the order of records in your own way, rather than having them be sortable on different attributes.

**Comment by seanlinsley** [comment]
It's implied in the ticket, but `attributes_table` needs to support `status_tag` in the same way that `table_for` does. As a user of AA, I really want my boolean fields to automatically use `status_tag` to display themselves.

**Comment by deivid-rodriguez** [comment]
@seanlinsley Do you know the status of this?

**Comment by javierjulio** [comment]
This will not be done as `table_for` and `attributes_table_for` should remain separate. If anything `attributes_table_for` will be phased out for a new `attributes_list_for` component (coming soon) that uses an HTML description list (formerly, definition list) that has a more flexible and responsive layout and styling then a `table` element.
